src/smt/smt_engine.cpp

```diff
@@ -4595,74 +4595,120 @@ Result SmtEngine::checkSynth(const Expr& e) throw(Exception) {
-  if( conj.getKind()==kind::FORALL ){
-    //possibly run quantifier elimination to make formula into single invocation
-    if( conj[1].getKind()==kind::EXISTS ){
-      Node conj_se = Node::fromExpr( expandDefinitions( conj[1][1].toExpr() ) );
-
-      Trace("smt-synth") << "Compute single invocation for " << conj_se << "..." << std::endl;
-      quantifiers::SingleInvocationPartition sip;
-      std::vector< Node > funcs;
-      funcs.insert(funcs.end(), conj[0].begin(), conj[0].end());
-      sip.init( funcs, conj_se );
-      Trace("smt-synth") << "...finished, got:" << std::endl;
-      sip.debugPrint("smt-synth");
-
-      if( !sip.isPurelySingleInvocation() && sip.isNonGroundSingleInvocation() ){
-        //We are in the case where our synthesis conjecture is exists f. forall xy. P( f( x ), x, y ), P does not contain f.
-        //The following will run QE on (exists z x.) exists y. P( z, x, y ) to obtain Q( z, x ),
-        //  and then constructs exists f. forall x. Q( f( x ), x ), where Q does not contain f.  We invoke synthesis solver on this result.
-
-        //create new smt engine to do quantifier elimination
-        SmtEngine smt_qe( d_exprManager );
-        smt_qe.setLogic(getLogicInfo());
-        Trace("smt-synth") << "Property is non-ground single invocation, run QE to obtain single invocation." << std::endl;
-        //partition variables
-        std::vector< Node > qe_vars;
-        std::vector< Node > nqe_vars;
-        for( unsigned i=0; i<sip.d_all_vars.size(); i++ ){
-          Node v = sip.d_all_vars[i];
-          if( std::find( sip.d_si_vars.begin(), sip.d_si_vars.end(), v )==sip.d_si_vars.end() ){
-            qe_vars.push_back( v );
-          }else{
-            nqe_vars.push_back( v );
+  if (options::sygusQePreproc())
+  {
+    // the following does quantifier elimination as a preprocess step
+    // for "non-ground single invocation synthesis conjectures":
+    //   exists f. forall xy. P[ f(x), x, y ]
+    // We run quantifier elimination:
+    //   exists y. P[ z, x, y ] ----> Q[ z, x ]
+    // Where we replace the original conjecture with:
+    //   exists f. forall x. Q[ f(x), x ]
+    // For more details, see Example 6 of Reynolds et al. SYNT 2017.
+    Node conj = Node::fromExpr(e);
+    if (conj.getKind() == kind::FORALL)
+    {
+      if (conj[1].getKind() == kind::EXISTS)
+      {
+        Node conj_se = Node::fromExpr(expandDefinitions(conj[1][1].toExpr()));
+
+        Trace("smt-synth") << "Compute single invocation for " << conj_se
+                           << "..." << std::endl;
+        quantifiers::SingleInvocationPartition sip;
+        std::vector<Node> funcs;
+        funcs.insert(funcs.end(), conj[0].begin(), conj[0].end());
+        sip.init(funcs, conj_se.negate());
+        Trace("smt-synth") << "...finished, got:" << std::endl;
+        sip.debugPrint("smt-synth");
+
+        if (!sip.isPurelySingleInvocation()
+            && sip.isNonGroundSingleInvocation())
+        {
```

Comment:
Instead of these chained ``if`` to do the transformation you could use them as conditions to give up and return, so that the code is not so nested.

Reply:
I'd like to keep only one call to checkSatisfiability in this method.  This would mean that I'd have to make a separate method that returns when it fails.  Unfortunately SMT engine is very big and disorganized as it is, so I don't think this is a good option at the moment unfortunately.  

A higher level question is whether this code belongs here, or whether we could move it to quantifiers, although this would require some more major changes.

Reply:
OK, opened issue #1363 for that.